lib/strings.nix

```diff
@@ -398,24 +406,8 @@ rec {
     let
       sep = addContextFrom _s _sep;
       s = addContextFrom _sep _s;
-      sepLen = stringLength sep;
-      sLen = stringLength s;
-      lastSearch = sLen - sepLen;
-      startWithSep = startAt:
-        substring startAt sepLen s == sep;
-
-      recurse = index: startAt:
-        let cutUntil = i: [(substring startAt (i - startAt) s)]; in
-        if index <= lastSearch then
-          if startWithSep index then
-            let restartAt = index + sepLen; in
-            cutUntil index ++ recurse restartAt restartAt
-          else
-            recurse (index + 1) startAt
-        else
-          cutUntil sLen;
     in
-      recurse 0 0;
+      builtins.filter builtins.isString (builtins.split (escapeRegex sep) s);
```

Comment:
This doesn't yet work with string context, which the previous version could handle:
```
nix-repl> lib.splitString "/" pkgs.hello                                
error: the string '/' is not allowed to refer to a store path (such as '!out!/nix/store/5rj29hp44c8p6281imxz9h1klqa23ijs-hello-2.10.drv'), at /home/infinisil/src/nixpkgs/lib/strings.nix:410:42
```

To fix this:
```diff
diff --git a/lib/strings.nix b/lib/strings.nix
index 9b702b57cb7..d81e46a1763 100644
--- a/lib/strings.nix
+++ b/lib/strings.nix
@@ -404,10 +404,11 @@ rec {
   */
   splitString = _sep: _s:
     let
-      sep = addContextFrom _s _sep;
-      s = addContextFrom _sep _s;
+      sep = builtins.unsafeDiscardStringContext _sep;
+      s = builtins.unsafeDiscardStringContext _s;
+      splits = builtins.filter builtins.isString (builtins.split (escapeRegex sep) s);
     in
-      builtins.filter builtins.isString (builtins.split (escapeRegex sep) s);
+      map (v: addContextFrom _sep (addContextFrom _s v)) splits;
 
   /* Return a string without the specified prefix, if the prefix matches.
```

This discards the string context of `s` and `sep` before using it, but adds their contexts back to all the parts again later.

A test for this would also be nice

Reply:
@Infinisil How would you test that the implementation keeps the context? Why use `unsafeDiscardStringContext _sep` instead of `"${_sep}"`?

Reply:
Ah hm, there is `builtins.getContext`, but it's not available in Nix 2.2 (which is what `lib/minver.nix` says). So maybe it's fine without tests for now.

If you try `"${_sep}"` you'll see that it doesn't work.